File: Backend/backend/src/main/java/com/paint/backend/Shapes/Shape.java
package com.paint.backend.Shapes;

import org.json.JSONObject;
import java.util.Objects;
import java.util.Stack;

public abstract class Shape implements IShape {
    protected String id;
    protected boolean draggable,strokeScaleEnabled;
    protected float x, y, strokeWidth , scaleX = 1, scaleY = 1;
    protected String className, stroke, fill;
    protected Stack<JSONObject> UndoUpdate = new Stack<>();
    protected Stack<JSONObject> RedoUpdate = new Stack<>();

    public void setID(String ID) { this.id = ID; }

    public void setType(String type) { this.className = type; }

    public void clearRedo() {
        RedoUpdate.clear();
    }

    public boolean isUpdated(){
        return !UndoUpdate.isEmpty();
    }
    public abstract JSONObject draw();

    public void delete(){
        UndoUpdate.push(new JSONObject().put("delete", true));
    }

    public String undoUpdate(){
        if(UndoUpdate.empty()){
            RedoUpdate.push(new JSONObject().put("create",true));
            return "delete";
        }else if(UndoUpdate.peek().has("delete")){
            RedoUpdate.push(UndoUpdate.pop());
            return "create";
        }else{
            update(UndoUpdate.pop(),"old");
        }
        return "update";
    }

    public String redoUpdate(){
        if(RedoUpdate.peek().has("create")){
            RedoUpdate.pop();
            return "create";
        }else if(RedoUpdate.peek().has("delete")){
            UndoUpdate.push(RedoUpdate.pop());
            return "delete";
        }else{
            update(RedoUpdate.pop(),"new");
        }
        return "update";
    }

    public void update(JSONObject jsonUpdate, String state){
        if(Objects.equals(state, "new")){ UndoUpdate.push(jsonUpdate); }
        else{ RedoUpdate.push(jsonUpdate); }

        System.out.println(jsonUpdate);
        switch(jsonUpdate.getString("key")){
            case "move"-> {
                this.x = (((JSONObject)(jsonUpdate.get(state))).getFloat("x"));
                this.y = (((JSONObject)(jsonUpdate.get(state))).getFloat("y"));
            }
            case "scale"-> {
                this.x = (((JSONObject)(jsonUpdate.get(state))).getFloat("x"));
                this.y = (((JSONObject)(jsonUpdate.get(state))).getFloat("y"));
                this.scaleX = (((JSONObject)(jsonUpdate.get(state))).getFloat("scaleX"));
                this.scaleY = (((JSONObject)(jsonUpdate.get(state))).getFloat("scaleY"));
            }
            case "style"->{
                this.strokeWidth = (((JSONObject)(jsonUpdate.get(state))).getFloat("strokeWidth"));
                this.stroke = (((JSONObject)(jsonUpdate.get(state))).getString("stroke"));
                this.fill = (((JSONObject)(jsonUpdate.get(state))).getString("fill"));
            }
        }
    }
}

File: frontend/src/app/Components/toolbar/shapeskit/shapes/Shape.ts
import Konva from "konva";
import { Ishape } from "./Ishape";

export abstract class Shape implements Ishape {
    protected _x?: number | undefined;
    protected _y?: number | undefined;
    protected _fill?: string | undefined;
    protected _stroke?: string | undefined;
    protected _strokeWidth?: number | undefined;
    protected _draggable?: boolean | undefined;
    protected _scaleX?: number | undefined;
    protected _scaleY?: number | undefined;

    constructor(x?:number, y?:number, fill?: string, stroke?: string, strokeWidth?: number, draggable?: boolean, scaleX?: number, scaleY?: number){
        this._x = x;
        this._y = y;
        this._fill = fill;
        this._stroke = stroke;
        this._strokeWidth = strokeWidth;
        this._draggable = draggable;
        this._scaleX = scaleX;
        this._scaleY = scaleY;
    }

    public get x(): number | undefined{
        return this._x;
    }
    public set x(value: number | undefined) {
        this._x = value;
    }
    public get y(): number | undefined{
        return this._y;
    }
    public set y(value: number | undefined) {
        this._y = value;
    }
    public get fill(): string | undefined{
        return this._fill;
    }
    public set fill(value: string | undefined) {
        this._fill = value;
    }
    public get stroke(): string | undefined {
        return this._stroke;
    }
    public set stroke(value: string | undefined) {
        this._stroke = value;
    }
    public get strokeWidth(): number | undefined {
        return this._strokeWidth;
    }
    public set strokeWidth(value: number | undefined) {
        this._strokeWidth = value;
    }
    public get draggable(): boolean | undefined {
        return this._draggable;
    }
    public set draggable(value: boolean | undefined) {
        this._draggable = value;
    }
    public get scaleX(): number | undefined {
        return this._scaleX;
    }
    public set scaleX(value: number | undefined) {
        this._scaleX = value;
    }
    public get scaleY(): number | undefined {
        return this._scaleY;
    }
    public set scaleY(value: number | undefined) {
        this._scaleY = value;
    }

    getKonva() {
        return new Konva.Shape({
            name: 'shape',
            x: this._x,
            y: this._y,
            fill: this._fill,
            stroke: this._stroke,
            strokeWidth: this._strokeWidth,
            draggable: this._draggable,
            scaleX: this._scaleX,
            scaleY: this._scaleY,
            strokeScaleEnabled: false,
        });
    }
}